src/sentry/sentry_metrics/indexer/ratelimiters.py

```diff
@@ -0,0 +1,151 @@
+        ] + [
+            Quota(prefix_override=None, **args)
+            for args in options.get("sentry-metrics.writes-limiter.limits.performance.per-org")
+        ]
+    elif use_case_id == UseCaseKey.RELEASE_HEALTH:
+        return [
+            Quota(prefix_override=_build_quota_key(UseCaseKey.RELEASE_HEALTH, None), **args)
+            for args in options.get("sentry-metrics.writes-limiter.limits.releasehealth.global")
+        ] + [
+            Quota(prefix_override=None, **args)
+            for args in options.get("sentry-metrics.writes-limiter.limits.releasehealth.per-org")
+        ]
+    else:
+        raise ValueError(use_case_id)
+
+
+@metrics.wraps("sentry_metrics.indexer.construct_quota_requests")
+def _construct_quota_requests(
+    use_case_id: UseCaseKey, keys: KeyCollection
+) -> Tuple[Sequence[OrgId], Sequence[RequestedQuota]]:
+    org_ids = []
+    requests = []
+    quotas = _construct_quotas(use_case_id)
+
+    if quotas:
+        for org_id, strings in keys.mapping.items():
+            org_ids.append(org_id)
+            requests.append(
+                RequestedQuota(
+                    prefix=_build_quota_key(use_case_id, org_id),
+                    requested=len(strings),
+                    quotas=quotas,
+                )
+            )
+
+    return org_ids, requests
+
+
+RateLimitState = Tuple[UseCaseKey, Sequence[RequestedQuota], Sequence[GrantedQuota], Timestamp]
+
+
+class WritesLimiter:
+    def __init__(self) -> None:
+        self.rate_limiters: MutableMapping[UseCaseKey, RedisSlidingWindowRateLimiter] = {}
+
+    def _get_rate_limiter(self, use_case_id: UseCaseKey) -> RedisSlidingWindowRateLimiter:
+        if use_case_id not in self.rate_limiters:
+            options = get_ingest_config(use_case_id).writes_limiter_cluster_options
+            self.rate_limiters[use_case_id] = RedisSlidingWindowRateLimiter(**options)
+
+        return self.rate_limiters[use_case_id]
+
+    @metrics.wraps("sentry_metrics.indexer.check_write_limits")
+    def check_write_limits(
+        self, use_case_id: UseCaseKey, keys: KeyCollection
+    ) -> Tuple[RateLimitState, KeyCollection, Sequence[Tuple[KeyResult, FetchType, FetchTypeExt]]]:
+        """
+        Takes a KeyCollection and applies DB write limits as configured via sentry.options.
+
+        Returns a tuple of:
+
+        1. `RateLimitState` that needs to be passed to `apply_write_limits` in order
+          to commit quotas upon successful DB write.
+
+        2. A key collection containing all unmapped keys that passed through the
+          rate limiter.
+
+        3. All unmapped keys that did not pass through the rate limiter.
+        """
+
+        org_ids, requests = _construct_quota_requests(use_case_id, keys)
+        timestamp, grants = self._get_rate_limiter(use_case_id).check_within_quotas(requests)
+
+        granted_key_collection = dict(keys.mapping)
```

Comment:
Why do you need to make a copy of the mapping ?

Also why is this called `granted_key_collection` ? This is the original mapping. It includes both allowed and rejected keys.

Reply:
it doesn't. in the last line of the for-loop we write back the list of strings that fit into the quota.